pkg/ebpf/net_events.go

```diff
@@ -25,8 +31,156 @@ type CaptureData struct {
+}
+
+func (tn *traceeNet) PutPcapWriter(id processPcapId, writer *pcapgo.NgWriter) {
+	tn.mtx.Lock()
+	defer tn.mtx.Unlock()
+
+	tn.pcapWriters[id] = writer
+}
+
+func (tn *traceeNet) DeletePcapWriter(id processPcapId) {
+	tn.mtx.Lock()
+	defer tn.mtx.Unlock()
+
+	delete(tn.pcapWriters, id)
+}
+
+type processPcapId struct {
+	hostPid       uint32
+	procStartTime uint64
+	comm          string
+	contID        string
+}
+
+func (t *Tracee) createPcapsDirPath(pcapContext processPcapId) (string, error) {
+
+	pcapsDirPath := path.Join(t.config.Capture.OutputPath, pcapContext.contID)
+	err := os.MkdirAll(pcapsDirPath, os.ModePerm)
+	if err != nil {
+		return "", err
+	}
+
+	return pcapsDirPath, nil
+}
+
+func (t *Tracee) getPcapFilePath(pcapContext processPcapId) (string, error) {
+
+	pcapsDirPath, err := t.createPcapsDirPath(pcapContext)
+	if err != nil {
+		return "", err
+	}
+
+	var pcapFileName string
+	if t.config.Capture.NetPerProcess {
+		pcapFileName = fmt.Sprintf("%s_%d_%d.pcap", pcapContext.comm, pcapContext.hostPid, pcapContext.procStartTime)
+	} else {
+		pcapFileName = "capture.pcap"
+	}
+
+	return path.Join(pcapsDirPath, pcapFileName), nil
+}
+
+func (t *Tracee) createPcapFile(pcapContext processPcapId) error {
+
+	pcapFilePath, err := t.getPcapFilePath(pcapContext)
+	if err != nil {
+		return fmt.Errorf("error getting pcap file path: %v", err)
+	}
+
+	pcapFile, err := os.OpenFile(pcapFilePath, os.O_APPEND|os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		return fmt.Errorf("error creating pcap file: %v", err)
+	}
+
+	ngIface := pcapgo.NgInterface{
+		Name:       t.config.Capture.NetIfaces[0],
+		Comment:    "tracee tc capture",
+		Filter:     "",
+		LinkType:   layers.LinkTypeEthernet,
+		SnapLength: uint32(math.MaxUint16),
+	}
+
+	pcapWriter, err := pcapgo.NewNgWriterInterface(pcapFile, ngIface, pcapgo.NgWriterOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, iface := range t.config.Capture.NetIfaces[1:] {
+		ngIface = pcapgo.NgInterface{
+			Name:       iface,
+			Comment:    "tracee tc capture",
+			Filter:     "",
+			LinkType:   layers.LinkTypeEthernet,
+			SnapLength: uint32(math.MaxUint16),
+		}
+
+		_, err := pcapWriter.AddInterface(ngIface)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Flush the header
+	err = pcapWriter.Flush()
+	if err != nil {
+		return err
+	}
+	t.netPcap.PutPcapWriter(pcapContext, pcapWriter)
+
+	return nil
+}
+
+func (t *Tracee) netExit(pcapContext processPcapId) {
+	// wait a second before deleting from the map - because there might be more packets coming in
+	time.Sleep(time.Second * 1)
+	t.netPcap.DeletePcapWriter(pcapContext)
+}
+
+func (t *Tracee) getPcapContext(hostTid uint32) (processPcapId, procinfo.ProcessCtx, error) {
+
+	networkThread, err := t.getProcessCtx(hostTid)
+	if err != nil {
+		return processPcapId{}, procinfo.ProcessCtx{}, err
+	}
+	networkProcess, err := t.getProcessCtx(networkThread.HostPid)
+	if err != nil {
+		return processPcapId{}, procinfo.ProcessCtx{}, err
+	}
```

Comment:
In case hostTid wasn't found, this error will repeat many times (for every packet), right?
This might cause error counter to have very big values, without understanding why exactly.
One thing I would change here is adding more context to the error, instead of just sending it as is, which might be unclear when trying to understand the problem.
Another thing to consider is to continue after sending the error (calling handleError directly instead of return)and just fill the missing fields with emty values. The advantage of doing that is that we save all the packets, even when there is an error of this kind, so we don't loose information about the network traffic